### ISSUE TITLE
508-defect-1 ❗ Launchblocker [SCREEN READER, COGNITION] Unable to update address, phone number and email address. Silent error messages.

### ISSUE THREAD

**Issue by AngelaFowler82** [department-of-veterans-affairs/va.gov-team#23252]
# 508-defect-1 :exclamation: Launchblocker

Issue Title: 508 Defect Level 1, Launch blocker, Screen readers, Cognitive Disabilities, unable to update address, phone number and email address. Silent error messages.

Point of Contact: Angela Fowler
**VFS Point of Contact:** _First name only: Angela

<h2>User Story or Problem Statement:</h2>
When accessing a profile page, I expect to be able to update basic information such as my address, phone number and email address when that functionality is promised. Also, as a screen reader user, I expect error messages to be plainly spoken and the means of correcting errors to be obvious.

<h2>Details</h2> 
<p> When updating mailing and home addresses, I entered an address in my home town I know doesn't exist. The system asked me to confirm the address, then informed me "We couldn't update your mailing address." In both cases the alerts weren't spoken and needed to be accessed by arrowing down from the level 3 heading "addresses." Also, when the notification appears on the screen, the prompt to edit my address remains, potentially causing confusion for users.</p>
<p> When I entered an address I knew existed, I was informed in the same way that "We are unable to update your mailing address. </p>
 <p>When updating my phone number, there is no indication as to the desired format. I tried entering the number three different ways: 
<UL>
<li>(XXX) XXX-XXXX</li>
<li>XXX-XX-XXXX</li>
<li>XXXXXXXXX</li>
</ul>
In all cases a notification appears stating "We could not update your phone number." This notification is not spoken by JAWS, and must be accessed by arrowing through the page. </p>
<p> When I tried to add an email address I got a message stating "We could not update your email address." This notification is not spoken by JAWS, and must be accessed by arrowing through the page. 
<p> When clicking from one section to another I.E. from address to phone number," if I did not cancel my attempt to change the previous section I got an error message stating that I must first finish my work in one section before proceeding to another. This notification is not spoken by JAWS and must be accessed by arrowing through the page.</p> 

<h2>Acceptance Criteria</h2>
- [ ] A user is able to successfully update their phone number,
- [ ] The desired number format, if any, is plainly stated,
- [ ] The user is able to successfully update their address,
- [ ] The user is able to successfully update their email,
- [ ]When it is necessary to confirm an address with the post office, that fact is indicated by a heading such as "confirmation required,"
- [ ] All error messages are obvious to all users,
- [ ] There is a stated method for correcting all errors.

<h2>Steps to Recreate</h2>
Log into Profile 2.0 using 
vets.gov.user+1@gmail.com

<h2>Proposed Solution (if known)</h2>
<ul>
<li>Error messages can be presented as Aria alerts, or as pop-up messages requiring the user to choose "ok" to continue.</li>
<li>When an address needs to be confirmed, a heading with a phrase such as "confirmation required" will draw a screen reader user's focus to the situation.</li>
<li> If the form requires that a phone number be written a certain way, the format can be depicted within the form label.</li>
<li>In the case of any error, focus should be placed at the point of the error.</li>
</ul>

**Comment by erikphansen** [comment]
@AngelaFowler82 It seems as though you got really lucky (or unlucky depending on your point of view) because there must have been legitimate problems with saving contact info to VA Profile when you found these issues.

You raised a number of issues in this ticket so I'll try to address them all:

First, users do _not_ need to enter phone numbers in any sort of special format which is why we don't have any guidance around that. The fact that you were unable to save a phone number had nothing to do with the formats you tried. There was just a server issue with saving the VA Profile when you were getting all of the errors.

Second, it looks like wrapping the save error alerts in a `div` with an `alert` role makes Voiceover read the alerts correctly. In order for you to validate this, there needs to be a server error when trying to save the contact info you are changing. One potential way to simulate this is to disconnect your computer from the internet, perhaps by disconnecting wifi, right before you click the UPDATE button.

Third, you mentioned that when a user needs to confirm their address the alert text telling them they need to confirm their address was not read. I have attempted to remedy that by placing keyboard focus on that alert box and that seems to work in my testing when using Voiceover in Safari.

I hope to have these changes on staging later today.

**Comment by erikphansen** [comment]
@AngelaFowler82 the error messages should now be read by screen readers on staging 🎉

**Comment by AngelaFowler82** [comment]
This looks a lot better. The only thing I'm seeing is, when I enter a nonsense address and select "use this address" at the prompt to confirm, JAWS speaks the message "We can't update your mailing address," focus is not placed on the error message.

**Comment by erikphansen** [comment]
I don't think I'm explicitly managing focus when either alert appears as a result of user input. I would think that doing so would be a problem because it would require more work on the user's part to get the focus where it needs to be to continue moving forward with their task (either confirming their address or attempting to save again if it failed the first time).

Could you provide guidance on how to manage focus when:
- a save error occurs when they enter a valid address
- the user is asked to confirm or validate their address
- a save error occurs when they are confirming their address

**Comment by AngelaFowler82** [comment]
I’ll have some more detailed guidelines when I come back on Wednesday, but as a user my first inclination would be to close out the error message since JAWS suggests I should do this. With focus placed on the message I can find the close button easily. With focus placed at the top, which it is when focus isn’t managed, it takes several seconds of searching to find the button as it is one of many buttons on the page.

**Comment by erikphansen** [comment]
@AngelaFowler82 What do you think about closing this particular issue? I think that the crux of this issue has been resolved since the error messages are now read by screen readers. The focus management issue is a slightly different problem that might be best tracked in a separate ticket especially because I need a little more guidance on how to best resolve it.

**Comment by AngelaFowler82** [comment]
Sorry for the delayed response. Let's go ahead and close. I'll write up the ticket addressing focus management.
